pkg/queue/breaker.go

```diff
@@ -99,60 +140,67 @@ type Semaphore struct {
 	mux      sync.Mutex
 }
 
-// Acquire receives the token from the semaphore, potentially blocking
+// Acquire receives the token from the semaphore, potentially blocking.
 func (s *Semaphore) Acquire() {
 	<-s.queue
 }
 
-// Release releases the token to the queue
-// The operation is potentially blocking when the queue is full
-func (s *Semaphore) Release() {
-	s.AddCapacity(1)
+// Release potentially puts the token back to the queue.
+// If the semaphore capacity was reduced in between and is not yet reflected,
+// we remove the tokens from the rotation instead of returning them back.
+func (s *Semaphore) Release() error {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	if s.reducers > 0 {
+		s.capacity--
+		s.reducers--
+	} else {
+		// We want to make sure releasing a token is always non-blocking.
+		select {
+		case s.queue <- s.token:
+		default:
+			// This should never happen.
+			return ErrRelease
+		}
+	}
+	return nil
 }
 
-// ReduceCapacity removes tokens from the rotation
-// It tries to acquire as many tokens as possible, if there are not enough tokens in the queue,
-// it postpones the operation for the future by increasing the `reducers` counter
-func (s *Semaphore) ReduceCapacity(size int32) error {
+// AddCapacity increases the number of tokens in the rotation.
+// Increases only if the size is positive, does nothing otherwise.
+// An error is returned if not all capacity could be added to the semaphore.
+// This error could happen when the number of tokens exceeds the semaphore's queue size.
+func (s *Semaphore) AddCapacity(size int32) error {
 	s.mux.Lock()
 	defer s.mux.Unlock()
-	if size > s.capacity {
-		return errors.New("the capacity that is released must be <= to added capacity")
-	}
 	for i := int32(0); i < size; i++ {
 		select {
-		case <-s.queue:
-			s.capacity--
+		case s.queue <- s.token:
+			s.capacity++
 		default:
-			s.reducers++
+			return ErrAddCapacity
 		}
 	}
 	return nil
 }
 
-// AddCapacity conditionally adds capacity to the semaphore
-// If there are tokens that must be reduced, release them first
-// Otherwise, add tokens to the queue
-func (s *Semaphore) AddCapacity(size int32) {
-	var leftToAdd int32
+// ReduceCapacity removes tokens from the rotation.
+// It tries to acquire as many tokens as possible, if there are not enough tokens in the queue,
+// it postpones the operation for the future by increasing the `reducers` counter.
+// We return an error when attempting to reduce more capacity then was originally added.
+func (s *Semaphore) ReduceCapacity(size int32) error {
 	s.mux.Lock()
 	defer s.mux.Unlock()
-	if s.reducers > 0 {
-		// do not allow reducers to be negative
-		if s.reducers >= size {
-			s.reducers -= size
-			s.capacity -= size
-		} else {
-			leftToAdd = size - s.reducers
-			s.reducers = 0
-		}
-	} else {
-		leftToAdd = size
+	if size > s.capacity || size+s.reducers > s.capacity {
```

Comment:
I think this check is redundant, `size+s.reducers > s.capacity` should suffice. If reducers are 0, the check is the same.